posts/running-background-tasks-python.md

```diff
@@ -0,0 +1,185 @@
+
+### Writing the parser
+
+Let's start with writing a simple parser that accepts a Goodreads book page URL. We will be using requests python library for making an HTTP request to get HTML content of the page. BeautifulSoup is a python library that lets us search, manipulate and create structured markup languages such as HTML, XML, etc. This will allow us to retrieve key information like the book title, author, rating, and description.
+
+```python
+# server.py
+
+def parse_book_link_for_meta_data(bookLink):
+  htmlString = requests.get(bookLink).content # Fetch HTML string of the book information page
+  bsTree = BeautifulSoup(htmlString,"html.parser") # Build a searchable tree using fetched HTML
+  # Find the required book attributes in the tree
+  title = bsTree.find("h1", attrs={"id": "bookTitle"}).string
+  author = bsTree.find("a", attrs={"class": "authorName"}).span.string
+  rating = bsTree.find("span", attrs={"itemprop": "ratingValue"}).string
+  description = ''.join(bsTree.find("div", attrs={"id": "description"}).find("span", attrs={"style": "display:none"}).stripped_strings)
+  return dict(title=title.strip() if title else '',author=author.strip() if author else '',rating=float(rating.strip() if rating else 0),description=description)
+```
+
+We can now write a function called `parse_and_persist_book_info` that calls the above parsing function and persists the value to Redis so that it can be retrieved later. This function along with its arguments will be pushed to queue so that the worker process can execute it. Redis is a key-value store where the key should be unique else it may lead to overwriting of a previous value. Here `generate_redis_key_for_book` is a function that generates a unique key for a given book URL.
+
+```python
+# server.py
+import pickle
+
+#........
+#........
+
+# This generates a unique Redis key against a book URL
+generate_redis_key_for_book = lambda bookURL: 'GOODREADS_BOOKS_INFO:' + bookURL
+
+def parse_and_persist_book_info(bookUrl):
+  redisKey = generate_redis_key_for_book(bookUrl) # get Redis key for given book URL 
+  bookInfo  = parse_book_link_for_meta_data(bookUrl) # get book meta information from parsing function above
+  # Set the value to Redis. Here pickle serializes the dictionary
+  redisClient.set(redisKey,pickle.dumps(bookInfo))
+```
+
+### The endpoint for accepting URLs
+
+Let's set up an endpoint that will accept a list of valid Goodreads book URLs. This is going to support POST method with URLs accepted as an array in _application/json_ body format. For validating the Goodreads book URLs we check for unique occurrences of URLs which starts with the string `https://www.goodreads.com/book/show/`. After the validation check, all valid URLs are pushed to Redis queue for parsing information. Here the method `enqueue_call` of Redis queue instance takes in a function that will be executed by worker process along with required arguments of the function.
```

Comment:
I would take out the first couple sentences, and make the paragraph this -->
After the validation check, all valid URLs are pushed to Redis queue for parsing information. Here the method `enqueue_call` of Redis queue instance takes in a function that will be executed by worker process along with required arguments of the function.

I don't think the first couple sentences add to the explanation of Redis Queues.